### PR TITLE
test(core): improve code coverage

### PR DESCRIPTION
**Pull request by arnaudbzn**
Add tests to improve code coverage for `./src/core`

Before:
Screenshot 2021-11-04 at 10 30 07

After:
Screenshot 2021-11-04 at 10 30 15

For some tests I had to "force" some use cases  by accessing private functions and props (I use `// Force....` in the test code comments. It means that we can maybe remove some pieces of existing code that are not necessary.

## Not covered
No solution found  to tests the following cases to reach 100% :pensive:
- [ ] `utils.scheduleMicrotask`: Mock `process.on` to test uncaught exception but issue with jest version support.
- [ ]  `utils.isPlainObject`: Find the corresponding object case!
- [ ]  `logger`:  Mock console to undefined or assume that the console is always defined and change the existing code `logger`definition.

## Impacts on existing code
Code changes to improve code coverage:
- [x]  `notifyManager.ts:` added a `notifyManagerTest` singleton for testing only
- [x] `queryCache.ts`: fixed comparison in`find()` filters object is always defined but could be empty.
- [x]  `queryObserver.ts`: removed useless condition  `shouldNotifyListeners`

### CODE REVIEW

File: src/core/notifyManager.ts [hunk 1]
```diff
@@ -94,3 +94,6 @@ class NotifyManager {
 // SINGLETON
 
 export const notifyManager = new NotifyManager()
+
+// For testing purposes (not change made by jest.setup.js)
+export const notifyManagerTest = new NotifyManager()
```
Comment (TkDodo) [inline src/core/notifyManager.ts:99]: can we tie this to `process.env.NODE_ENV !== 'production'`, or maybe even better, create the `new NotifyManager()` inside the test?

File: src/core/queryObserver.ts [hunk 2]
```diff
@@ -590,10 +590,6 @@ export class QueryObserver<
       return true
     }
 
-    if (result === prevResult) {
-      return false
-    }
```
Comment (TkDodo) [inline src/core/queryObserver.ts:595]: why was this removed please?

File: src/core/tests/focusManager.test.tsx [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    focusManager.setFocused(false)
+    expect(subscriptionSpy).toHaveBeenCalledTimes(0)
+
+    unsubscribe()
+  })
+
+  it('should return true for isFocused if document is undefined', async () => {
+    const { document } = globalThis
+
+    // @ts-ignore
```
Comment (TkDodo) [inline src/core/tests/focusManager.test.tsx:56]: let's please use `ts-expect-error` instead of `ts-ignore` wherever possible: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#-ts-expect-error-comments

File: src/core/tests/queryObserver.test.tsx [hunk 4]
```diff
@@ -490,4 +495,275 @@ describe('queryObserver', () => {
...
+
+    const spy = jest.fn()
+    const unsubscribe = observer.subscribe(spy)
+    await queryClient.refetchQueries(key)
+    await sleep(10)
+
+    // Force isStale to false
+    // because no use case has been found to reproduce this condition
+    // @ts-ignore
+    observer['currentResult'].isStale = true
```
Comment (TkDodo) [inline src/core/tests/queryObserver.test.tsx:762]: hmm, the `staleTime` is `20`, but we only wait for `10`. I think if we reverse this (set staleTime to 10 and wait for 20), `isStale` should automatically go to `true`.

**Review by TkDodo — CHANGES_REQUESTED** [review]
thank you for all the effort that goes into this 🚀

**Comment by TkDodo** [comment]
also, we just added codecov reporting. We should see a nice increase if you update your PR, and maybe we could even increase the expected coverage in the settings?

File: src/core/queryObserver.ts (hunk 2 above)
Comment (arnaudbzn) [reply src/core/queryObserver.ts:595]: `shouldNotifyListeners` is only used by `updateResult` that already does: 
```ts
     // Only notify if something has changed
    if (shallowEqualObjects(this.currentResult, prevResult)) {
      return
    }
```
before calling `shouldNotifyListeners`.
That why this condition was not in the code coverage so we can safely remove it or we can keep it and add a test  to directly  call `shouldNotifyListeners` with the appropriate params to cover this condition.

Comment (TkDodo) [reply src/core/queryObserver.ts:595]: right, I think `shouldNotifyListeners` should not make the extra equals check on result and previousResult, so this is fine 👍

File: src/core/tests/focusManager.test.tsx (hunk 3 above)
Comment (arnaudbzn) [reply src/core/tests/focusManager.test.tsx]: Done when possible in https://github.com/tannerlinsley/react-query/pull/2876/commits/ca22c597d27e0e4138f480f642ecbea374368b7c
Still have to use ` @ts-ignore ` to avoid `Unused '@ts-expect-error' directive.` in some cases.

File: src/core/notifyManager.ts (hunk 1 above)
Comment (arnaudbzn) [reply src/core/notifyManager.ts]: Done in https://github.com/tannerlinsley/react-query/pull/2876/commits/d3e40aff30fdb8d967c5e5c793e5b4b7757c6630
I had to export `NotifyManager` class. 
I think it's the best solution for testing.

File: src/core/tests/queryObserver.test.tsx (hunk 4 above)
Comment (arnaudbzn) [reply src/core/tests/queryObserver.test.tsx]: If I do this change the else in the block of code below is not covered anymore:


As mentioned in the test comment I have not found a use case for this condtion in the `setTimeout`:
https://github.com/tannerlinsley/react-query/blob/b52b2779b32b5cf0c21733e7d86b5882b8a611a6/src/core/queryObserver.ts#L364-L368

Comment (TkDodo) [reply src/core/tests/queryObserver.test.tsx]: I think what we'd need to do is manually set the query to stale by calling `queryClient.invalidateQueries(queryKey, { refetchActive: false })`. This would mark the query as `isInvalidated`, so `isStale` will already be true on the currentResult when the timer fires:

https://github.com/tannerlinsley/react-query/blob/e7a3207f74b27bc112d6ab583831cceac60f94c5/src/core/query.ts#L288-L294

Comment (arnaudbzn) [reply src/core/tests/queryObserver.test.tsx]: Tried but the `else` is still not covered and even the timeout handler is not called at all in this case.
It's because:  
`query.invalidate()`  => `dispatch()` => `onQueryUpdate()` => `updateTimers()` => `updateStaleTimeout()` => `clearStaleTimeout()`
That's why I don't see a use case for this code coverage.
It looks like a useless condition in the `setTimeout()` handler (or a bug smell?).

Comment (TkDodo) [reply src/core/tests/queryObserver.test.tsx]: True, it could be that the timer is reset once the query is marked as stale by invalidation. Maybe the condition is unnecessary but I’d rather not remove it now. Do you think it makes sense to keep the test?

Comment (arnaudbzn) [reply src/core/tests/queryObserver.test.tsx]: I was wondering the same for this kind of tests that test something that should never happen..
We can keep this test to improve the code coverage and I use the comment prefix `// Force ... ` to identify this kind of situation. 
So it makes sense to keep this test, it's a kind a code documentation about the tested condition.

**Comment by arnaudbzn** [comment]
@TkDodo It will be great if we can achieve 100% code coverage for the `./core` code base.
We can try do that in this PR or do you prefer to do that in another PR for the 3 remaining tests?

**Comment by TkDodo** [comment]
Let’s merge this one and do a follow up 👍

**Review by TkDodo — APPROVED** [review]
